include/pybind11/numpy.h

```diff
@@ -42,6 +43,46 @@ class array; // Forward declaration
 
 PYBIND11_NAMESPACE_BEGIN(detail)
 
+// Main author of this class: jbms@
+template <typename T>
+class LazyInitializeAtLeastOnceDestroyNever {
+public:
+    // PRECONDITION: The GIL must be held when `Get()` is called.
+    // It is possible that multiple threads execute `Get()` with `initialized_`
+    // still being false, and thus proceed to execute `initialize()`. This can
+    // happen if `initialize()` releases and reacquires the GIL internally.
+    // We accept this, and expect the operation to be both idempotent and cheap.
+    template <typename Initialize>
+    T &Get(Initialize &&initialize) {
+        if (!initialized_) {
+            assert(PyGILState_Check());
+            auto value = initialize();
+            if (!initialized_) {
+                new (reinterpret_cast<T *>(value_storage_)) T(std::move(value));
+                initialized_ = true;
+            }
+        }
+        PYBIND11_WARNING_PUSH
+#if !defined(__clang__) && defined(__GNUC__) && __GNUC__ < 5
+        // Needed for gcc 4.8.5
+        PYBIND11_WARNING_DISABLE_CLANG("-Wstrict-aliasing")
+#endif
+        return *reinterpret_cast<T *>(value_storage_);
+        PYBIND11_WARNING_POP
+    }
+
+    constexpr LazyInitializeAtLeastOnceDestroyNever() = default;
+#if __cplusplus >= 202002L
+    constexpr
+#endif // C++20
+        ~LazyInitializeAtLeastOnceDestroyNever()
+        = default;
```

Comment:
Can this go on a single line?

Reply:
Only with

```
// clang-format off
...
/// clang-format on
```

unfortunately. I feel it's better to just let clang-format do what it wants, but let me know if you prefer the override.

Reply:
Oh OK, sure, never mind!

Reply:
I found a nicer solution: `PYBIND11_DTOR_CONSTEXPR`